platform-modules/batch-models/src/main/scala/org/ekstep/analytics/job/StateAdminGeoReportJob.scala

```diff
@@ -0,0 +1,179 @@
+
+
+    val districtDF = subOrgJoinedDF.where(col("loctype").equalTo("district")).select(col("channel").as("channel"), col("slug"), col("id").as("schoolid"), col("orgname").as("schoolname"), col("locid").as("districtid"), col("locname").as("districtname"));
+    val blockDF = subOrgJoinedDF.where(col("loctype").equalTo("block")).select(col("id").as("schooljoinid"), col("locid").as("blockid"), col("locname").as("blockname"));
+
+    blockDF.join(districtDF, blockDF.col("schooljoinid").equalTo(districtDF.col("schoolid")), "left").drop(col("schooljoinid")).coalesce(1)
+      .select(col("schoolid").as("School id"),
+        col("schoolname").as("School name"),
+        col("channel").as("Channel"),
+        col("districtid").as("District id"),
+        col("districtname").as("District name"),
+        col("blockid").as("Block id"),
+        col("blockname").as("Block name"),
+        col("slug"))
+      .write
+      .partitionBy("slug")
+      .mode("overwrite")
+      .option("header", "true")
+      .csv(s"$detailDir")
+
+    fSFileUtils.renameReport(summaryDir, renamedDir, ".json", "geo-summary")
+    fSFileUtils.renameReport(detailDir, renamedDir, ".csv", "geo-detail")
+    fSFileUtils.purgeDirectory(detailDir)
+    fSFileUtils.purgeDirectory(summaryDir)
+
+     blockDF.distinct()
+  }
+
+  def saveGeoDetailsReport(reportDF: DataFrame, url: String): Unit = {
+    reportDF.coalesce(1)
+      .select(
+        col("id").as("School id"),
+        col("orgname").as("School name"),
+        col("channel").as("Channel"),
+        col("status").as("Status"),
+        col("locid").as("Location id"),
+        col("name").as("Location name"),
+        col("parentid").as("Parent location id"),
+        col("type").as("type"))
+      .write
+      .mode("overwrite")
+      .option("header", "true")
+      .csv(url)
+
+    JobLogger.log(s"StateAdminGeoReportJob: uploadedSuccess nRecords = ${reportDF.count()}")
+  }
+
+  /**
+   * Saves the raw data as a .json.
+   * Appends /summary to the URL to prevent overwrites.
+   * Check function definition for the exact column ordering.
+   * * If we don't partition, the reports get subsequently updated and we dont want so
+   * @param reportDF
+   * @param url
+   */
+  def saveSummaryReport(reportDF: DataFrame, url: String): Unit = {
+    reportDF.coalesce(1)
+      .write
+      .partitionBy("channel")
+      .mode("overwrite")
+      .json(url)
+
+    JobLogger.log(s"StateAdminGeoReportJob: uploadedSuccess nRecords = ${reportDF.count()}")
+    println(s"StateAdminGeoReportJob: uploadedSuccess nRecords = ${reportDF.count()} and ${url}")
+  }
+
+  def uploadReport(sourcePath: String) = {
+    // Container name can be generic - we dont want to create as many container as many reports
+    val container = AppConf.getConfig("cloud.container.reports")
+    val objectKey = AppConf.getConfig("admin.metrics.cloud.objectKey")
+
+    reportStorageService.upload(container, sourcePath, objectKey, isDirectory = Option(true))
```

Comment:
Use the OutputDispatcher - for ex: OutputDispatcher.dispatch("azure", Map("bucket" -> container, "file" -> sourcePath, "dir": Option(true)));